architecture/design/pessimistic-locking-functional-spec.md

```diff
@@ -0,0 +1,661 @@
+
+
+## 2 Requirements
+
+
+
+1. Firstly, support lock-lock type of conflict that only blocks, in YSQL. This is to split the current notion of conflict in YSQL into the two finer notions that Postgres supports.
+2. In case a transaction T1 tries to acquire a lock as part of a implicit/explicit lock acquisition that conflicts with existing locks held by active txns -
+    1. It has to wait for all txns that have conflicting locks to end (waiting behaviour)
+    2. Throw a serialization error only if a modification that conflicts with the lock has occurred.
+    3. Ability to timeout and remove self from queue
+3. A [distributed deadlock detection algorithm](https://docs.google.com/document/d/1E4LHGmVZuTlr36_uczPuE6aAjoLT4sfr1o0YCtzUUPc/edit#) to break cycles. The following properties are required -
+    4. No false positives
+    5. Bound on latency = O (cycle size)
+4. Add a metric to measure the “intensity” of starvation by measuring the number instances where a transaction jumps the wait queue ahead of other waiting transactions that it conflicts with just because it doesn’t conflict with active transactions.
+
+Once pessimistic locking is supported, YSQL will have the same behaviour as Postgres with regards to transaction waiting behaviour in case of writing/locking, with **_two exceptions_**:
+
+
+
+1. For Serializable isolation level, there will be an extra YSQL-only behaviour of waiting in case a read (or write) intent is to be written that conflicts with an already existing write (or read) intent from another transaction. In Pg’s Serializable isolation implementation (i.e., SSI), reads don’t take implicit locks and hence don’t “lock” conflict with other writes. And hence, a read doesn’t wait on another write (and vice versa) in Pg’s Serializable isolation.
+2. YSQL writes fine grained column level intents in case of modifications to specific columns only. This will allow us to be more fine grained so that modifications to different columns of a row need not result in waiting (possibly followed by a serialization error). This is one aspect in which YSQL would turn out to be **_better than_** Pg - and semantically different in a hopefully beneficial way.
+    1. In case customers are relying on this more coarse-grained locking behavior of PostgreSQL, we can’t help them without significant overhaul of existing behavior
+
+
+## 3 Surface Area & Usage
+
+
+#### Surface Area
+
+**yb_use_pessimistic_locking**: a cluster level gflag to turn on pessimistic locking. This will require a cluster restart. Note that all txns in the cluster will either use pessimistic locking or optimistic locking.
+
+
+#### Expected Behavior
+
+
+```
+create table test (k int primary key, v int);
+insert into test values (1, 1);
+```
+
+
+
+1. **Explicit lock - Explicit lock interaction (only blocks)**
+
+<table>
+  <tr>
+   <td>
+<code>> begin transaction isolation level repeatable read; \
```

Comment:
This seems like it won't render right with the two angle brackets...

Reply:
done